# 画像Data の加工・可視化
#### 利用Sean
- 画像認識による AI Model 開発
- 物体検知
- Motion Capture

## 画像File を読み込んで表示してみよう

In [10]:
from cv2 import cv2

img = cv2.imread('data/chapter5/sample.jpg')  # Image data の読み込み
img

array([[[ 8, 10,  4],
        [ 8, 10,  4],
        [ 8, 10,  4],
        ...,
        [47, 69, 45],
        [46, 66, 47],
        [45, 65, 46]],

       [[ 8, 10,  4],
        [ 8, 10,  4],
        [ 8, 10,  4],
        ...,
        [50, 72, 48],
        [49, 70, 48],
        [48, 69, 47]],

       [[ 9, 12,  3],
        [ 9, 12,  3],
        [ 9, 12,  3],
        ...,
        [52, 74, 49],
        [50, 72, 48],
        [48, 70, 46]],

       ...,

       [[20, 48,  5],
        [17, 47,  4],
        [15, 46,  1],
        ...,
        [27, 32, 31],
        [21, 21, 27],
        [15, 14, 23]],

       [[18, 49,  4],
        [17, 48,  3],
        [15, 47,  0],
        ...,
        [27, 31, 25],
        [20, 22, 22],
        [17, 16, 18]],

       [[18, 50,  3],
        [16, 48,  1],
        [12, 46,  0],
        ...,
        [26, 32, 21],
        [21, 22, 18],
        [17, 17, 17]]], dtype=uint8)

単純に img を出力しただけでは、数字の羅列が出力される。
> 読み込みを行なうことで、Computer が扱える形の数字Data に変換されているため。

In [11]:
# 画像を表示してみる
cv2.imshow('sample', img)
cv2.waitKey(0)

-1

## 画像Data の中身を確認しよう

In [12]:
img.shape

(3456, 5184, 3)

- 3456, 5184 は解像度（画像の大きさ、高さ x 横幅）を表す。
  > 一般的には、「横幅 x 高さ」表記が多いので要注意。
- 3 は色の数が赤、青、緑の３色で表現されているため、３チャンネルという意味になる。
  > Monochrome の場合は、1 になる。
  > 一般的には RGB という順番だが、OpenCV は、BGR の逆順になっている。

もう少し細かく Data を見るため、０番目の Data を指定する。

In [13]:
print(img[0])
print(img[0].shape)

[[ 8 10  4]
 [ 8 10  4]
 [ 8 10  4]
 ...
 [47 69 45]
 [46 66 47]
 [45 65 46]]
(5184, 3)


Data形状は、5184 x 3 となっていて、つまり０番目の高さの Data を取得しており、最初の\[8 10 14] はそれぞれ BGR の色情報を示している。

２つ目の横幅を 0 に指定してみる

In [14]:
print(img[:, 0])
print(img[:, 0].shape)

[[ 8 10  4]
 [ 8 10  4]
 [ 9 12  3]
 ...
 [20 48  5]
 [18 49  4]
 [18 50  3]]
(3456, 3)


前述と同様に \[8 10 4] のように BGR の Data が確認できる。
Data 形状は 3456 x 3 となっており、高さ x 色情報になっている。

色情報を固定してみる

In [15]:
print(img[:,:,0])  # 最後の Data のみ 0 にして、今回でいうと青色の Data のみ取得
print(img[:,:,0].shape)

[[ 8  8  8 ... 47 46 45]
 [ 8  8  8 ... 50 49 48]
 [ 9  9  9 ... 52 50 48]
 ...
 [20 17 15 ... 27 21 15]
 [18 17 15 ... 27 20 17]
 [18 16 12 ... 26 21 17]]
(3456, 5184)


- \[8 8 8 ... 47 46 45] の部分が5184個になり横幅方向の Data.
- [] で囲われた array が34356個ある。

画像Data は非構造Data ではあるが読み込んでしまえば、numpy の array として使用でき、数字Data として扱える。

## 画像Data を切り出してみよう

In [16]:
img_extract = img[700:1200, 300:800, :]  # 700~1199までの500個、300~799までの500個の 500x500　Data で Slice
cv2.imshow('Img Extract', img_extract)
cv2.waitKey(0)

-1

画像全体の左上の数字８が取得できている。

Color 指定をしてみる

In [17]:
cv2.imshow('Img Extract Blue', img_extract[:, :, 0])  # 青色読み込み
cv2.waitKey(0)

-1

全体的に黒色で画像が表示され、特に芝生の部分も黒く塗られてしまっている印象。

In [18]:
cv2.imshow('Img Extract Green', img_extract[:, :, 1])  # 緑色読み込み
cv2.waitKey(0)

-1

芝生の部分がかなり白色で表現されている。
Monochrome の場合、大きいほ値ほど白色になるため、今回は、緑色の要素が強い芝生が白くなっている。